server/sonar-server/src/main/java/org/sonar/server/issue/ws/SearchResponseLoader.java

```diff
@@ -87,9 +96,107 @@ public SearchResponseData load(Collector collector, @Nullable Facets facets) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      SearchResponseData result = new SearchResponseData(loadIssues(preloadedResponseData, collector, dbSession));
+      collector.collect(result.getIssues());
+
+      loadRules(preloadedResponseData, collector, dbSession, result);
+      // order is important - loading of comments complete the list of users: loadComments() is
+      // before loadUsers()
+      loadComments(collector, dbSession, result);
+      loadUsers(preloadedResponseData, collector, dbSession, result);
+      loadComponents(preloadedResponseData, collector, dbSession, result);
+      loadOrganizations(dbSession, result);
+      loadActionsAndTransitions(collector, result);
+      completeTotalEffortFromFacet(facets, result);
+      return result;
+    }
+  }
+
+  private List<IssueDto> loadIssues(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession) {
+    List<IssueDto> preloadedIssues = preloadedResponseData.getIssues();
+    Set<String> preloadedIssueKeys = preloadedIssues.stream().map(IssueDto::getKey).collect(MoreCollectors.toSet(preloadedIssues.size()));
+    Set<String> issueKeysToLoad = copyOf(difference(collector.getIssueKeys(), preloadedIssueKeys));
+    if (issueKeysToLoad.isEmpty()) {
+      return preloadedIssues;
+    }
+
+    List<IssueDto> loadedIssues = dbClient.issueDao().selectByOrderedKeys(dbSession, issueKeysToLoad);
+    return concat(preloadedIssues.stream(), loadedIssues.stream())
+      .collect(toList(preloadedIssues.size() + loadedIssues.size()));
+  }
+
+  private void loadUsers(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    if (collector.contains(USERS)) {
+      List<UserDto> preloadedUsers = firstNonNull(preloadedResponseData.getUsers(), emptyList());
+      Set<String> preloadedLogins = preloadedUsers.stream().map(UserDto::getLogin).collect(MoreCollectors.toSet(preloadedUsers.size()));
+      Set<String> requestedLogins = collector.get(USERS);
+      Set<String> loginsToLoad = copyOf(difference(requestedLogins, preloadedLogins));
+
+      if (loginsToLoad.isEmpty()) {
+        result.setUsers(preloadedUsers);
+      } else {
+        List<UserDto> loadedUsers = dbClient.userDao().selectByLogins(dbSession, loginsToLoad);
+        result.setUsers(concat(preloadedUsers.stream(), loadedUsers.stream()).collect(toList(preloadedUsers.size() + loadedUsers.size())));
+      }
+    }
+  }
+
+  private void loadComponents(SearchResponseData preloadedResponseData, Collector collector, DbSession dbSession, SearchResponseData result) {
+    Collection<ComponentDto> preloadedComponents = preloadedResponseData.getComponents();
+    Set<String> preloadedComponentUuids = preloadedComponents.stream().map(ComponentDto::uuid).collect(MoreCollectors.toSet(preloadedComponents.size()));
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') A "NullPointerException" could be thrown; "preloadedComponents" is nullable here. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2259)